### DISCUSSION TITLE
How to derive `FromForm` on a struct which has `Option<ObjectId>`

### DISCUSSION THREAD

**Discussion by CyanFroste** [rwf2/Rocket#2314] [Questions] [answered]
![image](https://user-images.githubusercontent.com/53912961/183414050-c3ad71c1-1290-43e1-b914-11779f2481dd.png)

# Code

```rust

#[derive(Debug, Serialize, Deserialize, FromForm)]
#[serde(crate = "rocket::serde")]
struct Item {
    // https://github.com/thedodd/wither/issues/62
    #[serde(
        rename = "_id",
        skip_serializing_if = "Option::is_none",
        serialize_with = "serialize_object_id"
    )]
    id: Option<ObjectId>,
    name: String,
}

pub fn serialize_object_id<S: Serializer>(
    object_id: &Option<ObjectId>,
    serializer: S,
) -> Result<S::Ok, S::Error> {
    match object_id {
        Some(ref object_id) => serializer.serialize_some(&object_id.to_hex()),
        None => serializer.serialize_none(),
    }
}


#[post("/create", data = "<item>")]
async fn create(conn: Connection<Db>, item: Form<Item>) {
    let collection = conn
        .default_database()
        .expect("Db to exist")
        .collection::<Item>("items");

    let docs = vec![Item {
        id: None,
        name: "Okay".to_string(),
    }];

    collection
        .insert_many(docs, None)
        .await
        .expect("Items to be added");
}

```

**Comment by edgarogh** [comment, marked as answer]
You cannot derive `FromForm` if a member doesn't implement it. Because `Option<ObjectId>` doesn't belong to you, you can't derive `FromForm` on it either.

The typical Rusty way to deal with this is to create a wrapping newtype around `ObjectId` and [manually derive `FromFormField`](https://docs.rs/rocket/0.5.0-rc.2/rocket/form/trait.FromFormField.html#implementing) on it. You can use [`ObjectId::with_string`](https://docs.rs/mongodb/0.3.12/i686-pc-windows-msvc/mongodb/oid/struct.ObjectId.html#method.with_string) to build your Mongo ObjectId.

```rust
struct RocketObjectId(mongodb::bson::oid::ObjectId);

impl<'v> FromFormField<'v> for RocketObjectId {
    ...
}
```

**Reply by CyanFroste** [reply]
But then, I simply can't directly use it with the db... Thanks for the reply though.

**Reply by edgarogh** [reply]
Why? Because it can't be `serde::`(de)serialized for the MongoDB driver?

You can derive serde's traits on your newtype struct and add [`#[serde(transparent)]`](https://serde.rs/container-attrs.html#transparent) so the newtype structure (de)serializes exactly like its only field does.

Or you can also have two different structs — one for the DB and one for Rocket. That's a bit _WET_ but at the same time, it follows the principle of Separation of Concerns as it decouples the public API from the internal representation.

Or similarly, use a generic `MyDocument<IdType>` that you parameterize differently depending on which trait you need implemented at any point in time (if `FromForm` supports deriving on generic structures).
